### ISSUE TITLE
Network outputs only PADs

### ISSUE THREAD

**Issue by ishalyminov** [farizrahman4u/seq2seq#133]
Hi all,

I'm trying to put together a proof-of-concept tiny network and train it on a tiny part of OpenSubtitles dataset (200 question-answer pairs). I use a small vocabulary of 5000 words (I use pre-trained Google News word2vec embeddings for those), and all my utterances fit the smallest bucket (max_input_len=5, max_output_len=10), e.g.:

> PAD PAD PAD PAD neat; GO i like it out here STOP PAD PAD PAD

I was hoping to see the network memorize all the sequences, but after running 10, 100, 1000 epochs with each architecture (SimpleSeq2Seq, Seq2Seq, AttentionSeq2Seq), it outputs exclusively PAD symbols, with top-1 probability being really low (like 0.0015).

Can you please suggest how I can debug it further?

**Comment by phdowling** [comment]
Share your model code - I'm using the models on a QA task and things are working fine, training eventually seems to take off

**Comment by ishalyminov** [comment]
Here it is: https://github.com/ishalyminov/seq2seq_dialogue/blob/master/training_utils.py#L18

**Comment by phdowling** [comment]
I guess your last layer should be a TimeDistributed Softmax, no? I'm not sure how keras handles this internally. I've also had more success using RMSprop or another adaptive algorithm to train, perhaps that will help.

Perhaps you could share some of your hyperparameters (vectors, hidden layer/encoder/decoder numbers and sizes, sequence input/output lengths...)? Perhaps you are underfitting the problem.

**Comment by ishalyminov** [comment]
Wow, didn't know about the TimeDistributedSoftmax, thanks! Although I tried without the Softmax layer at all (and MSE objective) - the same basic behavior..
My parameters are here: https://github.com/ishalyminov/seq2seq_dialogue/blob/master/sanity_check.json

**Comment by ishalyminov** [comment]
I actually don't see time_distributed_softmax among the Keras activations: https://github.com/fchollet/keras/blob/master/keras/activations.py#L6 - maybe softmax already works with time distributed data (it's mentioned there that it can handle 3D data)?

**Comment by ishalyminov** [comment]
Could you describe how you pass data into the network and what you input/output tensors look like (also, is there any difference if I set batch_input_shape or input_shape, output_shape individually)?
And do you use both encoder and decoder input for training, as in this trick: https://github.com/farizrahman4u/seq2seq/blob/master/examples/train.py#L25 (not even sure how it works - @farizrahman4u can you please explain?)

**Comment by phdowling** [comment]
@ishalyminov Yeah, I guess the softmax is fine (in any case, you would use the TimeDistributed() wrapper around your Activation layer). Have you changes your optimizer yet? It can have a huge impact on training performance. Just try `"rmsprop"`, it will use default settings, this has usually worked for me. Your learning rate seems large to me, but I guess it is in line with Keras defaults..

**Comment by farizrahman4u** [comment]
@phdowling  Can you write an example for seq2seq? Will mention credits on readme.

**Comment by phdowling** [comment]
@farizrahman4u I'm currently using it for character level QA stuff, which might be a slightly untypical use case - How about simple character copy toy task, i.e. recognizing opening and closing tags in a sequence of random chars, and copying what's inside the tasks? I have some code for this, and a relatively simple network can solve this with near 100% accuracy.

**Comment by farizrahman4u** [comment]
Sounds cool
